backend/src/Edna/Web/Types.hs

```diff
@@ -17,4 +43,180 @@ data ExperimentalMeasurement = ExperimentalMeasurement
+data WithId t = WithId
+  { wiId :: SqlId t
+  , wItem :: t
+  } deriving stock (Generic, Show)
+
+-- | This data type is used when you want to return something with its ID and
+-- some additional data that was not submitted by end users, but is maintained
+-- by the application.
+data WithExtra t e = WithExtra
+  { weId :: SqlId t
+  , weItem :: t
+  , weExtra :: e
+  } deriving stock (Generic, Show)
+
+-- | A stub to specify the sorting order, most likely will be replaced with
+-- @servant-util@.
+data StubSortBy =
+    SortByName
+  | SortBySomething
+
+instance FromHttpApiData StubSortBy where
+  parseQueryParam = \case
+    "name" -> pure SortByName
+    "something" -> pure SortBySomething
+    x -> Left $ "unknown sorting order: " <> x
+
+----------------
+-- Entities
+----------------
+
+-- | Input data submitted along with uploaded file.
+data FileUploadInput = FileUploadInput
+  { fuiProject :: SqlId Project
+  -- ^ ID of the project the file belongs to.
+  , fuiTestMethodology :: SqlId TestMethodology
+  -- ^ ID of the test methodology used throughout the file.
+  } deriving stock (Generic, Show)
+
+-- | Summary of an experiment data file.
+newtype FileSummary = FileSummary
+  { unFileSummary :: [FileSummaryItem]
+  } deriving stock (Generic, Show)
+
+-- | One element in 'FileSummary'. Corresponds to one target from the file.
+-- Contains all compounds that interact with the target in the file.
+-- Also contains information whether this target is new or already known.
+data FileSummaryItem = FileSummaryItem
+  { fsiTarget :: SqlId Target
+  -- ID of a target from the file.
+  , fsiCompounds :: [SqlId Compound]
+  -- IDs of all compounds interacting with this target.
+  , fsiIsNew :: Bool
+  -- Whether this target is new, i. e. wasn't uploaded before.
+  } deriving stock (Generic, Show)
+
+-- | Project as submitted by end users.
+data Project = Project
+  { pName :: Text
+  , pDescription :: Text
+  } deriving stock (Generic, Show)
+
+-- | Extra data about projects that is not submitted by users, but is stored
+-- in DB.
+data ProjectExtra = ProjectExtra
+  { peCreationDate :: Word64
+  -- ^ Unsigned number, UNIX timestamp.
+  -- Can be changed to a more specific type later.
+  -- The same applies to other timestamps.
+  , peLastUpdate :: Word64
+  , peCompoundNames :: [Text]
+  -- ^ Names of all compounds involved in this project.
+  } deriving stock (Generic, Show)
+
+-- | Test methodology as submitted by end users.
+data TestMethodology = TestMethodology
+  { tmName :: Text
+  , tmDescription :: Text
+  , tmConfluence :: URI
+  } deriving stock (Generic, Show)
+
+-- | Compounds are not submitted directly by users, so for now
+-- there is only one representation for frontend.
+-- MDe links are trivial to generate, so we offload this task to frontend.
+data Compound = Compound
+  { cName :: Text
+  -- ^ Name of the compound, it may be changed to be a number later.
+  , cChemSoft :: Maybe URI
+  -- ^ Link to ChemSoft.
```

Comment:
No `CreationDate`

Reply:
(just randomly saw this comment)

it's not needed here, compounds are not something what is created manually but automatically when a new file uploaded and our system is not a source of truth, so it can't be creation date but rather date when it was automatically added to our system.

Reply:
This type corresponds directly to this table from the wireframe:
![image](https://user-images.githubusercontent.com/2556409/111278376-8317a980-864a-11eb-96f7-066513d1a768.png)
 Currently there is no creation date there, so we don't return it. Though I see that we store it in DB, so we can return and let frontend decide what to do with it (in current design it can ignore this field).

Anyway, @pva701 and @sashasashasasha151, I want to point out that the set of tables in wireframes is really _tentative_, we just put some stuff that we find reasonable there, but something will likely change there when the customer starts testing our software. So for now let's just use our existing wireframe as the source of truth and implement everything according to it. AFAIU there is nothing very complex, but we can amend it if we heavily dislike something. I am also talking about @pva701's comment about showing projects for targets. We don't know what will should be shown for Targets, so we just picked some data that _probably_ makes sense. AFAIU we can get all projects by some SQL query, so it shouldn't be a big deal, though probably a bit bigger deal than returning some small data that's directly available.

Regarding `CreationDate`, I agree that there is inconsistency: I added it to `init.sql` because I thought that _maybe this information will be useful_, but at the same time I didn't add it here because it's currently not in the wireframe. I think I'll just add it to `Compound` to avoid inconsistency and frontend will ignore it for now.

Regarding naming, yes, compounds are not created manually, so it's a date of _automatic_ creation of an entry in our database.

Reply:
>Anyway, @pva701 and @sashasashasasha151, I want to point out that the set of tables in wireframes is really tentative, we just put some stuff that we find reasonable there, but something will likely change there when the customer starts testing our software.

But targets have it

Reply:
Yes, and I guess that's one of the reasons why I put `creationDate` for `compound` in SQL. Anyway, as you can see, I've added `cAdditionDate :: Word64` to `Compound`, so I assume the problem is solved, isn't it?